[#118479861] Seed and rotate RDS master password

saliceti opened the pull request:
# What

Story: [Generating secure master password for Postgres Instances](https://www.pivotaltracker.com/story/show/118479861)

We want to add a seed to generate the DB instances master passwords so they can't easily be guessed. The seed is set using a broker property.
As a consequence, we need to be able to rotate the passwords when the seed is changed. This is implemented by checking the connection on all instances at startup and resetting the password if necessary.
This relies on tags being applied on broker created instances. A temporary task was added to the pipeline to tag existing instances.

This should be tested with PRs on paas-aws-broker-boshrelease, paas-cf and paas-docker-cloudfoundry-tools.
# How to review
- Deploy on an existing environment. It should detect the RDS passwords have changed and reset the passwords.
- Check the rds broker logs to confirm
- Bind an existing DB
- Create and bind a new DB
- Unbind
- Run autodelete
### Expected error when the password is wrong

This should no happen:

```
$ cf bind-service mtest healthcheck-db
Binding service healthcheck-db to app mtest in org testers / space healthcheck as admin...
FAILED
Server error, status code: 502, error code: 10001, message: Service broker error: pq: password authentication failed for user "DJ4YDF83aJXWxX26"
```
# Who can review

Anyone but @saliceti, @mtekel, @keymon

keymon committed:
Read broker_name from configuration

Read from the configuration the name of the current broker, which
will be used to tag the RDS DB instances.

Tagging the instances will later allow us to map them to specific
broker deployments.

keymon committed:
Tag the instances with the Broker Name

Add a tag "Broker Name" with the name passed in the configuration.

keymon committed:
Allow describe instances by Tag

Implement a method to list all the existing instances which have a
given tag with a given value.

This will allow us retrieve from AWS all the RDS instances with a specific
tag "Broker Name", and then check if they require reset the master
password.

saliceti committed:
Add seed and broker_name to configuration doc

saliceti committed:
Update sample config with seed and broker_name

keymon committed:
Use STS instead of IAM to retrieve account ID

IAM.GetUser() doesn't work with instance profile. We replace it with
STS.GetCallerIdentity() that works both locally and with instance
profile.

saliceti committed:
Update sample IAM policy

We don't need to allow GetUser anymore but we need ListTagsForResource.

saliceti committed:
Re-add password seed changes

This reverts commit da2a651109be7387c7334559c566bea5c2466b27.

These changes were reverted so as to unblock other changes on master,
not because the changes themselves were unwanted. This therefore
re-instates these changes.

saliceti committed:
Replace IAM dependencies with STS

Update AWS SDK to latest version and add missing AWS packages

combor commented on rdsbroker/config_test.go:
```diff
@@ -61,6 +62,14 @@ var _ = Describe("Config", func() {
 			Expect(err.Error()).To(ContainSubstring("Must provide a non-empty MasterPasswordSeed"))
 		})
 
+		It("returns error if BrokerName is not valid", func() {
```
"returns error if BrokerName is empty" ?

combor commented on rdsbroker/broker.go:
```diff
@@ -391,10 +391,57 @@ func (b *RDSBroker) LastOperation(instanceID string) (brokerapi.LastOperationRes
+
+		var dbName string
+		if dbInstanceDetails.DBName != "" {
+			dbName = dbInstanceDetails.DBName
```
Why this can be empty, maybe it's better to use `b.dbName(dbInstanceDetails.Identifier)` always ?

saliceti replied:
The [rest of the code](https://github.com/alphagov/paas-rds-broker/blob/73d2b88af73894de0767fe36192716d836034ffb/rdsbroker/broker.go#L235-L239) always assume this can be empty. I think it's safer to leave it like this as it may not be reliable. An RDS instance may not hold any databases for example.

combor commented on rdsbroker/broker.go:
```diff
@@ -391,10 +391,57 @@ func (b *RDSBroker) LastOperation(instanceID string) (brokerapi.LastOperationRes
+				return err
+			}
+		} else {
+			defer sqlEngine.Close()
```
It would be good to move this just after `sqlEngine.Open()` so it's more readable that we care about closing connections.

combor commented on CONFIGURATION.md:
```diff
@@ -9,8 +9,13 @@ A sample configuration can be found at [config-sample.json](https://github.com/a
 | username   | Y        | String | Broker Auth Username
 | password   | Y        | String | Broker Auth Password
+| master_password_seed | Y        | String | Seed to generate DB instances master passwords
+| broker_name | Y        | String | Unique name to identify by tag in RDS the instances created by this broker
```
It's a little bit unclear description. I would change it to sth like "Unique name of RDS broker, used to construct a tag for instance identification"

mtekel replied:
I have updated `broker_name` property description in spec in https://github.com/alphagov/paas-aws-broker-boshrelease/pull/8 to "Unique name of RDS broker, used to construct a tag for instance identification". I think we should use same string here...

combor commented on sqlengine/mysql_engine.go:
```diff
@@ -34,6 +34,11 @@ func (d *MySQLEngine) Open(address string, port int64, dbname string, username s
 }
 
+func (d *MySQLEngine) Ping() error {
+	panic("Not implemented")
```
Why we want to panic here ?

saliceti replied:
Because we should never come here.
We would implement it when we look at MySQL.

combor replied:
Why not returning nil? It will continue to work like it's implemented without panicking.

saliceti replied:
Because that would hide the error. You would think that it pings successfully when it fact it does nothing.

alext replied:
Why can't we do a `d.db.Ping()` here as well? It's a function defined by Go's SQL package, not by a specific driver.

alext replied:
Alternatively, if we don't want to implement this for MySQL at this point, it could simply return `errors.New("Not Implemented")`.

The problem with using `panic` here is that it will cause the broker to exit. This means that if a mysql instance gets deployed for any reason, the broker will cease to function at all until this code is updated.

dcarley commented on rdsbroker/broker.go:
```diff
@@ -387,20 +391,79 @@ func (b *RDSBroker) LastOperation(instanceID string) (brokerapi.LastOperationRes
+		err = sqlEngine.Open(dbAddress, dbPort, dbName, masterUsername, b.masterPassword(serviceInstanceID))
+		if err != nil {
+			b.logger.Error("Error opening connection to the database", err)
+			return err
```
I'm unsure about returning from this loop if we get an error on a single instance. It would mean that if all passwords need rotating but one instance is down then some instances could be left with the incorrect password.

To change that though we'd need to either return `[]error` or split the code within the loop out to a separate function that we call/log multiple times.

What do you think?

saliceti replied:
I could call `continue` and go on with the other instances?

dcarley replied:
Perhaps.. how would we know about the failures?

saliceti replied:
I would log the error and continue

dcarley commented on rdsbroker/broker.go:
```diff
@@ -387,20 +391,79 @@ func (b *RDSBroker) LastOperation(instanceID string) (brokerapi.LastOperationRes
+func (b *RDSBroker) CheckAndRotateCredentials() error {
+	dbInstanceDetailsList, err := b.dbInstance.DescribeByTag("Broker Name", b.brokerName)
+	if err != nil {
+		b.logger.Error("Could not retrieve the list of DB instances", err)
```
I think all of these errors will get logged twice, once within this function and once from the caller in `main.go`, which could be quite confusing. Should we remove the log statements within this function and let the caller handle reporting?

saliceti replied:
Should we do the opposite? These logs are more specific and can help to understand the issue.
I could remove the log from main.go instead.

dcarley replied:
That doesn't feel right to me. I think the consumer of the function (either `main()` or the test suite) should know and report errors.

I agree that it can be useful to have additional context, but I suspect the kind of errors we'll get back from this function will give us enough hints about where they've come from.

combor commented on rdsbroker/broker.go:
```diff
@@ -387,20 +391,79 @@ func (b *RDSBroker) LastOperation(instanceID string) (brokerapi.LastOperationRes
 	return lastOperationResponse, nil
 }
 
+func (b *RDSBroker) CheckAndRotateCredentials() error {
```
It would be good to have a discussion whether we want to bundle checking and rotation of credentials in one function. Personally I would go for splitting this to two functions. `CheckCredentials` that can return list of instances that it can't connect to and `RotateCredentials` that can consume the list and change passwords. It makes those functions more generic and re-usable.

dcarley commented on rdsbroker/broker.go:
```diff
@@ -387,20 +391,79 @@ func (b *RDSBroker) LastOperation(instanceID string) (brokerapi.LastOperationRes
+		err = sqlEngine.Ping()
+		if err != nil {
+			if strings.Contains(err.Error(), "Login failed") {
+				b.logger.Error(fmt.Sprintf("Login failed when connecting to DB %v at %v, will attempt to reset the password", dbName, dbAddress), err)
```
I think this should be an "info" log message rather than "error" because it's an expected action if we've rotated the seed.

dcarley commented on rdsbroker/broker.go:
```diff
@@ -387,20 +391,79 @@ func (b *RDSBroker) LastOperation(instanceID string) (brokerapi.LastOperationRes
+			b.logger.Error("Error opening connection to the database", err)
+			return err
+		} else {
+			defer sqlEngine.Close()
```
Minor: can you put the `defer` outside the `if` condition? We're going to continue at this point anyway, so it doesn't need to be conditional and it's a bit easier to read.

combor replied:
> can you put the defer outside the if condition

Github somehow hides my comment. Even if you click on the link you have to scroll up to find it:
https://github.com/alphagov/paas-rds-broker/pull/11/commits/80200ad2c6fb28052254c31495e3ac988d0eb95f#r67339023

dcarley commented on rdsbroker/broker_test.go:
```diff
@@ -1703,4 +1709,125 @@ var _ = Describe("RDS Broker", func() {
+
+				It("should not try to change the master password", func() {
+					err := rdsBroker.CheckAndRotateCredentials()
+					Expect(err).To(HaveOccurred())
```
Can we be more specific about the error we're expecting?

saliceti replied:
I don't know what we're expecting but I know we're not expecting login failed so I can check that.

dcarley commented on rdsbroker/broker_test.go:
```diff
@@ -1703,4 +1709,125 @@ var _ = Describe("RDS Broker", func() {
+					},
+				}
+
+				sqlEngine.PingError = errors.New("Login failed")
```
It's not clear to me from the error checking in the test, or the descriptions of the `Context` and `It`, why this error is here. Could you explain?

saliceti replied:
"We reset the password..." is not clear enough?

dcarley commented on rdsbroker/broker_test.go:
```diff
@@ -1703,4 +1709,125 @@ var _ = Describe("RDS Broker", func() {
+				sqlEngine.PingError = errors.New("Login failed")
+			})
+
+			BeforeEach(func() {
```
Could this be combined with the other `BeforeEach` so that it's easier to read?

saliceti replied:
Ok. They initialise different things so I thought that would make it clearer. Ob... Not.

alext commented on awsrds/rds_db_instance_test.go:
```diff
@@ -610,10 +722,10 @@ var _ = Describe("RDS DB Instance", func() {
+			getCallerIdentityInput *sts.GetCallerIdentityInput
+			getCallerIdentityError error
 		)
+		account := "123456789012"
```
It's not good to assign this in this way. Written like this, it only gets initialised once, and if any test modifies it, that change will persist to other tests.

If this value never changes, then we should define it as a `const`. Otherwise, it should be defined like the other variables here (with the variable declared in the `var` block above, and initialized in the `BeforeEach`).

alext commented on sqlengine/postgres_engine.go:
```diff
@@ -37,6 +40,18 @@ func (d *PostgresEngine) Open(address string, port int64, dbname string, usernam
+	err := d.db.Ping()
+	if err != nil {
+		d.logger.Error("sql-ping-error", err)
+		if strings.Contains(err.Error(), "password authentication failed") {
```
It's not good practice to use string matching to check errors. It would be better to use a [type assertion](https://golang.org/ref/spec#Type_assertions) to cast the error to a `*pq.Error`, and then check the error code directly.

alext commented on sqlengine/postgres_engine.go:
```diff
@@ -37,6 +40,18 @@ func (d *PostgresEngine) Open(address string, port int64, dbname string, usernam
+	if err != nil {
+		d.logger.Error("sql-ping-error", err)
+		if strings.Contains(err.Error(), "password authentication failed") {
+			return errors.New("Login failed")
```
It would be better to define a package-level variable for this error, and return that. That way a caller can test against that error instead of having to do string matching.

saliceti committed:
DRY retrieving DBName

The code was repeated in 3 different places, it is now in one method.

saliceti committed:
Ping the connection inside Postgres engine Open()

Open() may not actually connect to the database so we are not sure
the connection details are right:
https://golang.org/pkg/database/sql/#Open

It is an issue as we rely on it to validate the connection details.
We now ping the connection to validate properly.

A new generic error LoginFailedError is returned in case of login
failure.

saliceti committed:
Add CheckAndRotateCredentials function

At startup time, check connection to all DB instances with the known
master password and reset the password in case of failure.

This allows to rotate the passwords automatically after changing the
seed.

alext commented on CONFIGURATION.md:
```diff
@@ -9,8 +9,13 @@ A sample configuration can be found at [config-sample.json](https://github.com/a
 | username   | Y        | String | Broker Auth Username
 | password   | Y        | String | Broker Auth Password
+| master_password_seed | Y        | String | Seed to generate DB instances master passwords
+| broker_name | Y        | String | RDS broker name used to tag instances for identification
```
These values aren't in the top-level config block, they're under the `rds_config` section below.

alext commented on main.go:
```diff
@@ -62,14 +62,20 @@ func main() {
 
 	serviceBroker := rdsbroker.New(config.RDSConfig, dbInstance, sqlProvider, logger)
 
+	err = serviceBroker.CheckAndRotateCredentials()
```
I think this should run in a separate goroutine so that it doesn't delay startup of the broker. As it stands, startup will take an ever increasing amount of time as the number of RDS instances grows.

combor commented on main.go:
```diff
@@ -62,14 +62,17 @@ func main() {
 
 	serviceBroker := rdsbroker.New(config.RDSConfig, dbInstance, sqlProvider, logger)
 
+	go serviceBroker.CheckAndRotateCredentials()
```
I'm unsure about error handling here. If we care about returning errors we should something like this:

```
go func() {
  err := serviceBroker.CheckAndRotateCredentials()
  if err != nil {
     logger.Error(err)
  }
}
```

If we want to log errors inside CheckAndRotateCredentials let's do not return them.

combor commented on rdsbroker/broker.go:
```diff
@@ -387,20 +383,70 @@ func (b *RDSBroker) LastOperation(instanceID string) (brokerapi.LastOperationRes
+		return
+	}
+
+	b.logger.Debug(fmt.Sprintf("Found %v RDS instances managed by the broker", len(dbInstanceDetailsList)))
```
Minor: `%d` instead of `%v` ?
